src/Eventso.Subscription.Kafka.DeadLetter.Postgres/PoisonEventStore.cs

```diff
@@ -0,0 +1,386 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Confluent.Kafka;
+using Eventso.Subscription.Kafka.DeadLetter.Store;
+using Npgsql;
+
+namespace Eventso.Subscription.Kafka.DeadLetter.Postgres
+{
+    public sealed class PoisonEventStore : IPoisonEventStore
+    {
+        private readonly IConnectionFactory _connectionFactory;
+        private readonly int _maxAllowedFailureCount;
+        private readonly TimeSpan _minIntervalBetweenRetries;
+
+        public PoisonEventStore(
+            IConnectionFactory connectionFactory,
+            // TODO receive from settings
+            int? maxAllowedFailureCount = default,
+            TimeSpan? minIntervalBetweenRetries = default)
+        {
+            _connectionFactory = connectionFactory;
+            _maxAllowedFailureCount = maxAllowedFailureCount ?? 10;
+            _minIntervalBetweenRetries = minIntervalBetweenRetries ?? TimeSpan.FromMinutes(1);
+        }
+
+        public static async Task<PoisonEventStore> Initialize(
+            IConnectionFactory connectionFactory,
+            // TODO receive from settings
+            int? maxAllowedFailureCount = default,
+            TimeSpan? minIntervalBetweenRetries = default)
+        {
+            await using var connection = connectionFactory.ReadWrite();
+            
+            await using var command = new NpgsqlCommand(@"
+                CREATE SCHEMA IF NOT EXISTS eventso_dlq;
+
+                CREATE TABLE IF NOT EXISTS eventso_dlq.poison_events (
+                    topic                  TEXT         NOT NULL,
+                    partition              INT          NOT NULL,
+                    ""offset""               BIGINT       NOT NULL,
+                    key                    UUID         NOT NULL,
+                    value                  BYTEA        NULL,
+                    creation_timestamp     TIMESTAMP    NOT NULL,
+                    header_keys            TEXT[]       NULL,
+                    header_values          BYTEA[]      NULL,
+                    last_failure_timestamp TIMESTAMP    NOT NULL,
+                    last_failure_reason    TEXT         NOT NULL,
+                    total_failure_count    INT          NOT NULL,
+                    PRIMARY KEY (""offset"", partition, topic)
+                );
+
+                CREATE INDEX IF NOT EXISTS ix_poison_events_topic ON eventso_dlq.poison_events (topic);",
```

Comment:
this index looks like useless. You can just reorder PK and move topic on the first place

Reply:
index by key now